# NumPy, Pandas, MLP

In [21]:
import numpy as np
from numpy import linalg
import pandas as pd

### Задание 1
Создайте numpy array с элементами от числа N до 0 (например, для N = 10 это будет array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0]))

### Решение

In [2]:
n = int(input())

matrix = np.arange(n, -1, -1)
print(type(matrix))
print(matrix)

5
<class 'numpy.ndarray'>
[5 4 3 2 1 0]


### Задание 2
Создайте диагональную матрицу с элементами от N до 0. Посчитайте сумму ее значений на диагонали.

### Решение

In [4]:
n = int(input())

matrix_1 = np.diag(np.flip(np.arange(0, n+1)), k = 0)
print(matrix_1)

10
[[10  0  0  0  0  0  0  0  0  0  0]
 [ 0  9  0  0  0  0  0  0  0  0  0]
 [ 0  0  8  0  0  0  0  0  0  0  0]
 [ 0  0  0  7  0  0  0  0  0  0  0]
 [ 0  0  0  0  6  0  0  0  0  0  0]
 [ 0  0  0  0  0  5  0  0  0  0  0]
 [ 0  0  0  0  0  0  4  0  0  0  0]
 [ 0  0  0  0  0  0  0  3  0  0  0]
 [ 0  0  0  0  0  0  0  0  2  0  0]
 [ 0  0  0  0  0  0  0  0  0  1  0]
 [ 0  0  0  0  0  0  0  0  0  0  0]]


In [5]:
matrix_1.sum()

55

In [6]:
np.trace(matrix_1)

55

### Задание 3
Решите систему уравнений:

4x + 2y + z = 4

x + 3y = 12

5y + 4z = -3

### Решение

In [10]:
a = np.array([[4, 2, 1], [1, 3, 0], [0, 5, 4]])
b = np.array([4, 12, -3])

In [11]:
# решение 1
print(linalg.solve(a, b))
print(np.dot(a, b))

[ 0.46666667  3.84444444 -5.55555556]
[37 40 48]


In [12]:
np.allclose(np.dot(a, linalg.solve(a, b)), b)

True

In [13]:
# решение 2
from scipy.linalg import solve

solve(a, b)

array([ 0.46666667,  3.84444444, -5.55555556])

In [14]:
print(np.allclose(np.dot(a, linalg.solve(a, b)), b))
print(np.allclose(np.dot(a, solve(a, b)), b))

True
True


### Задание 4

Имеется матрица покупок в интернет-магазине. Столбец А - ID пользователя. Остальные столбцы - количество покупок категорий товаров этим пользователем:

Матрица в виде numpy array

In [16]:
users_stats = np.array(
    [
        [2, 1, 0, 0, 0, 0],
        [1, 1, 2, 1, 0, 0],
        [2, 0, 1, 0, 0, 0],
        [1, 1, 2, 1, 0, 1],
        [0, 0, 1, 2, 0, 0],
        [0, 0, 0, 0, 0, 5],
        [1, 0, 0, 0, 0, 0],
        [0, 1, 1, 0, 0, 0],
        [0, 0, 0, 1, 1, 3],
        [1, 0, 0, 2, 1, 4]
    ], 
    np.int32
)

На сайт заходит очередной посетитель, о покупках которого известно следующее:

In [17]:
next_user_stats = np.array([0, 1, 2, 0, 0, 0])

Найдите самого похожего пользователя. Т. е. посчитайте косинусное сходство между этим пользователем и всеми пользователями из массива user_stats

### Решение

In [18]:
dists = []
lines = len(users_stats)
i = 0
for i in range(lines):
    """
    Подсчет косинуса угла между векторами a, b по их координатам
    """
    # длины векторов
    aLength = np.linalg.norm(users_stats[i])
    bLength = np.linalg.norm(next_user_stats)
    cosine = round(np.dot(users_stats[i], next_user_stats) / (aLength * bLength), 2)
    dists.append({"index_user": i, "distance": float(cosine), "vector": users_stats[i]}) 
    """
    создание результирующщего словаря из которого будет найдено max значение косинусного сходства между user'ами
    """
    i += 1

In [19]:
dists

[{'index_user': 0, 'distance': 0.2, 'vector': array([2, 1, 0, 0, 0, 0])},
 {'index_user': 1, 'distance': 0.85, 'vector': array([1, 1, 2, 1, 0, 0])},
 {'index_user': 2, 'distance': 0.4, 'vector': array([2, 0, 1, 0, 0, 0])},
 {'index_user': 3, 'distance': 0.79, 'vector': array([1, 1, 2, 1, 0, 1])},
 {'index_user': 4, 'distance': 0.4, 'vector': array([0, 0, 1, 2, 0, 0])},
 {'index_user': 5, 'distance': 0.0, 'vector': array([0, 0, 0, 0, 0, 5])},
 {'index_user': 6, 'distance': 0.0, 'vector': array([1, 0, 0, 0, 0, 0])},
 {'index_user': 7, 'distance': 0.95, 'vector': array([0, 1, 1, 0, 0, 0])},
 {'index_user': 8, 'distance': 0.0, 'vector': array([0, 0, 0, 1, 1, 3])},
 {'index_user': 9, 'distance': 0.0, 'vector': array([1, 0, 0, 2, 1, 4])}]

In [20]:
from scipy.spatial.distance import cosine

dists = []
lines = len(users_stats)
i = 0
for i in range(lines):
    dists.append({"index_user": i, "distance": round(cosine(users_stats[i], next_user_stats), 2)})
    i += 1
    
dists

[{'index_user': 0, 'distance': 0.8},
 {'index_user': 1, 'distance': 0.15},
 {'index_user': 2, 'distance': 0.6},
 {'index_user': 3, 'distance': 0.21},
 {'index_user': 4, 'distance': 0.6},
 {'index_user': 5, 'distance': 1.0},
 {'index_user': 6, 'distance': 1.0},
 {'index_user': 7, 'distance': 0.05},
 {'index_user': 8, 'distance': 1.0},
 {'index_user': 9, 'distance': 1.0}]

Ответ: юзер с индесом 7, вектор [0, 1, 1, 0, 0, 0]

In [22]:
def cosine(users_stats, next_user_stats):
    aLength = np.linalg.norm(users_stats, axis = 1)
    bLength = np.linalg.norm(next_user_stats)
    cosine = np.dot(users_stats, next_user_stats) / (aLength * bLength)
    return cosine

In [23]:
cosine(users_stats, next_user_stats)

array([0.2       , 0.84515425, 0.4       , 0.79056942, 0.4       ,
       0.        , 0.        , 0.9486833 , 0.        , 0.        ])

In [24]:
np.dot(users_stats, next_user_stats)

array([1, 5, 2, 5, 2, 0, 0, 3, 0, 0])

In [25]:
(aLength * bLength)

10.488088481701517

### Задание 5

Скачайте с сайта https://grouplens.org/datasets/movielens/ датасет любого размера. Определите какому фильму было выставлено больше всего оценок 5.0.

### Решение

In [26]:
# читаю данные с оценками юзеров и фильтруем данные оставляя только фильмы с оценкой 5.0

df_rating = pd.read_csv('C:/Netology/ml-25m/ratings.csv')
df_rating = df_rating.query('rating == 5.0')
df_rating.head()

userId  movieId  rating   timestamp
0        1      296     5.0  1147880044
2        1      307     5.0  1147868828
3        1      665     5.0  1147878820
8        1     1237     5.0  1147868839
18       1     2632     5.0  1147878248

In [27]:
# читаю данные с названиями и жанрами фильмов

df_movies = pd.read_csv('C:/Netology/ml-25m/movies.csv')
df_movies.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [28]:
# добавляю название фильма к данным с оценками юзеров

df_rating = df_rating.merge(df_movies[['title', 'movieId']], on='movieId', how='left')
df_rating.head()

userId  movieId  rating   timestamp  \
0       1      296     5.0  1147880044   
1       1      307     5.0  1147868828   
2       1      665     5.0  1147878820   
3       1     1237     5.0  1147868839   
4       1     2632     5.0  1147878248   

                                               title  
0                                Pulp Fiction (1994)  
1   Three Colors: Blue (Trois couleurs: Bleu) (1993)  
2                                 Underground (1995)  
3    Seventh Seal, The (Sjunde inseglet, Det) (1957)  
4  Saragossa Manuscript, The (Rekopis znaleziony ...

In [29]:
# применяю фильтр к оценкам, оставляем оценки равные 5.0 , применяем метод value_counts() и выводим 1 значение

df_rating[['title', 'rating']].value_counts().head(1)

title                             rating
Shawshank Redemption, The (1994)  5.0       39553
dtype: int64

## Ответ 
Фильм с ID равному 318 выставлено больше всего оценок 5.0, фильм Shawshank Redemption .

### Задание 6
По данным файла power.csv посчитайте суммарное потребление стран Прибалтики (Латвия, Литва и Эстония) категорий 4, 12 и 21 за период с 2005 по 2010 года. Не учитывайте в расчетах отрицательные значения quantity.

### Решение 

In [31]:
# читаем данные

countries = pd.read_csv('C:/Netology/Lecture_notebook/10._Basics_Pandas/power.csv')
countries.head()

country  year  quantity  category
0  Austria  1996       5.0         1
1  Austria  1995      17.0         1
2  Belgium  2014       0.0         1
3  Belgium  2013       0.0         1
4  Belgium  2012      35.0         1

In [32]:
# фильтруем и выбираем данные только по Латвии, Литве и Эстонии

pribaltica_countries = countries.query('country == "Latvia" or country == "Lithuania" or country == "Estonia"')
pribaltica_countries.head()

country  year  quantity  category
373  Lithuania  2014      39.0         1
374  Lithuania  2013     192.0         1
375  Lithuania  2012     196.0         1
376  Lithuania  2011     171.0         1
377  Lithuania  2010     158.0         1

In [33]:
# фильтр датафрейма стран Прибалтики с учетом категорий, вывод только категорий 4, 12 и 21
# фильтр датафрейма стран Прибалтики с учетом положительного значение quantity (quantity = 0 или quantity > 0)
# фильтр датафрейма стран Прибалтики с года (с 2005 по 2010 включительно)

filtered_baltics_countries = pribaltica_countries.query('category == 4 or category == 12 or category == 21').query('2005 <= year <= 2010 and quantity >= 0')
filtered_baltics_countries.head(5)

country  year  quantity  category
10617    Estonia  2008       1.0         4
10618    Estonia  2007       0.0         4
10619    Estonia  2006       0.0         4
10620    Estonia  2005       1.0         4
11156  Lithuania  2008       1.0         4

In [34]:
# суммирование столбца quantity
# группировка по столбцам country, year, category
# сортировка по year, country, category

filtered_baltics_countries.groupby(by=['country','year', 'category']).sum().sort_values(by=['year', 'country', 'category'], ascending=[True, True, True]).head(60)

quantity
country   year category          
Estonia   2005 4             16.0
               21         10236.0
Latvia    2005 21          8664.0
Lithuania 2005 4              6.0
               12            18.0
               21         18224.0
Estonia   2006 4              6.0
               21         10228.0
Latvia    2006 21          8604.0
Lithuania 2006 4              6.0
               12            18.0
               21         19868.0
Estonia   2007 4              0.0
               21         10508.0
Latvia    2007 4              0.0
               21          8532.0
Lithuania 2007 4              0.0
               12          2454.0
               21         19972.0
Estonia   2008 4              6.0
               21         10776.0
Latvia    2008 4              0.0
               21          8620.0
Lithuania 2008 4              6.0
               12          2029.0
               21         20140.0
Estonia   2009 21         10660.0
Latvia    2009 4              0.0
               21         10104.0
Lithuania 2009 12           862.0
               21         21156.0
Estonia   2010 21         11004.0
Latvia    2010 4              0.0
               21         10320.0
Lithuania 2010 12           945.0
               21         16592.0

### Задание 7
Выберите страницу любого сайта с табличными данными. Импортируйте таблицы в pandas dataframe.
Примеры страниц (необязательно брать именно эти):

### Решение

In [35]:
data = pd.read_html('https://www.bbc.com/russian/news-51706538')
data[0][0:20]

Unnamed: 0   Умерли  Смертность*  Заразились  \
0              США  600 464       1836.0  33 438 579   
1         Бразилия  509 141       2431.0  18 243 483   
2            Индия  393 310        291.0  30 134 445   
3          Мексика  232 068       1839.0   2 493 087   
4             Перу  191 073       5973.0   2 036 449   
5           Россия  129 869        891.0   5 346 005   
6   Великобритания  128 048       1907.0   4 684 572   
7           Италия  127 362       2101.0   4 255 700   
8          Франция  110 906       1706.0   5 764 329   
9         Колумбия  102 636       2067.0   4 060 013   
10       Аргентина   91 438       2061.0   4 350 564   
11        Германия   90 687       1091.0   3 732 969   
12            Иран   83 473       1020.0   3 140 129   
13         Испания   80 766       1730.0   3 777 539   
14          Польша   74 953       1977.0   2 879 470   
15             ЮАР   59 406       1028.0   1 877 143   
16       Индонезия   55 949        209.0   2 053 995   
17         Украина   54 429       1230.0   2 295 873   
18          Турция   49 417        600.0   5 393 248   
19         Румыния   32 771       1680.0   1 080 457   

    Новые случаи  0  10  100  1000  10000  **  Unnamed: 5  
0                                         NaN         NaN  
1                                         NaN         NaN  
2                                         NaN         NaN  
3                                         NaN         NaN  
4                                         NaN         NaN  
5                                         NaN         NaN  
6                                         NaN         NaN  
7                                         NaN         NaN  
8                                         NaN         NaN  
9                                         NaN         NaN  
10                                        NaN         NaN  
11                                        NaN         NaN  
12                                        NaN         NaN  
13                                        NaN         NaN  
14                                        NaN         NaN  
15                                        NaN         NaN  
16                                        NaN         NaN  
17                                        NaN         NaN  
18                                        NaN         NaN  
19                                        NaN         NaN

In [36]:
data[0][0:2]

Unnamed: 0   Умерли  Смертность*  Заразились  \
0        США  600 464       1836.0  33 438 579   
1   Бразилия  509 141       2431.0  18 243 483   

   Новые случаи  0  10  100  1000  10000  **  Unnamed: 5  
0                                        NaN         NaN  
1                                        NaN         NaN

In [38]:
data[0][0:10]

Unnamed: 0   Умерли  Смертность*  Заразились  \
0             США  600 464       1836.0  33 438 579   
1        Бразилия  509 141       2431.0  18 243 483   
2           Индия  393 310        291.0  30 134 445   
3         Мексика  232 068       1839.0   2 493 087   
4            Перу  191 073       5973.0   2 036 449   
5          Россия  129 869        891.0   5 346 005   
6  Великобритания  128 048       1907.0   4 684 572   
7          Италия  127 362       2101.0   4 255 700   
8         Франция  110 906       1706.0   5 764 329   
9        Колумбия  102 636       2067.0   4 060 013   

   Новые случаи  0  10  100  1000  10000  **  Unnamed: 5  
0                                        NaN         NaN  
1                                        NaN         NaN  
2                                        NaN         NaN  
3                                        NaN         NaN  
4                                        NaN         NaN  
5                                        NaN         NaN  
6                                        NaN         NaN  
7                                        NaN         NaN  
8                                        NaN         NaN  
9                                        NaN         NaN

### Задание 8

Задание 1**
Напишите функцию, которая классифицирует фильмы из материалов занятия по следующим правилам:
- оценка 2 и меньше - низкий рейтинг
- оценка 4 и меньше - средний рейтинг
- оценка 4.5 и 5 - высокий рейтинг

Результат классификации запишите в столбец class

### Решение

In [39]:
# открываем и читаем таблицу с рейтингами

ratings = pd.read_csv('../ml-25m/ratings.csv')
ratings.head()

userId  movieId  rating   timestamp
0       1      296     5.0  1147880044
1       1      306     3.5  1147868817
2       1      307     5.0  1147868828
3       1      665     5.0  1147878820
4       1      899     3.5  1147868510

In [40]:
# считаем средний рейтинг для каждого фильма и формируем таблицу со средним рейтингом

average_rating = ratings[['movieId','rating']].groupby('movieId').mean().round(2)
average_rating.head()

rating
movieId        
1          3.89
2          3.25
3          3.14
4          2.85
5          3.06

In [41]:
# функция присваивающая классификацию фильма на основе его рейтинг

def class_movie(rating):
    if rating <= 2:
        return 'Низкий рейтинг'
    elif rating > 2 and rating <= 4:
        return  'Средний рейтинг'
    elif rating == 4.5 or rating == 5:
        return 'Высокий рейтинг'

In [42]:
# применяем функцию к датафрейму со средним рейтигом фильмов

average_rating['class'] = average_rating['rating'].apply(class_movie)
average_rating.head()

rating            class
movieId                         
1          3.89  Средний рейтинг
2          3.25  Средний рейтинг
3          3.14  Средний рейтинг
4          2.85  Средний рейтинг
5          3.06  Средний рейтинг

In [43]:
average_rating.sample(5)

rating            class
movieId                         
163775     2.86  Средний рейтинг
165791     3.00  Средний рейтинг
161216     2.75  Средний рейтинг
193641     3.00  Средний рейтинг
115381     3.17  Средний рейтинг

### Задание 9

Используем файл keywords.csv.

Необходимо написать гео-классификатор, который каждой строке сможет выставить географическую принадлежность определенному региону. Т. е. если поисковый запрос содержит название города региона, то в столбце ‘region’ пишется название этого региона. Если поисковый запрос не содержит названия города, то ставим ‘undefined’.

Правила распределения по регионам Центр, Северо-Запад и Дальний Восток:

In [44]:
geo_data = {
'Центр': ['москва', 'тула', 'ярославль'],
'Северо-Запад': ['петербург', 'псков', 'мурманск'],
'Дальний Восток': ['владивосток', 'сахалин', 'хабаровск']
}

### Решение

In [45]:
# читаем наш файл

keywords = pd.read_csv('C:/Netology/Python_11_pandas_functions_groupby/keywords.csv')
keywords.head()

keyword     shows
0             вк  64292779
1  одноклассники  63810309
2          порно  41747114
3           ютуб  39995567
4      вконтакте  21014195

In [46]:
# функция которая ищет название города в запросе пользователя

def geo_classifier(keyword):
    for key, value in geo_data.items(): #используя цикл для того чтобы добраться до названия города и присвоеть его переменной City
        for city in value:
            if city in keyword.lower().split(' '):
                return key        
    return 'undefined'

In [47]:
# применяем функцию к датафрейму с поисковыми запросами

keywords['region'] = keywords['keyword'].apply(geo_classifier)
keywords.sample(10)

keyword   shows     region
2315                                 ифнс   91536  undefined
60457          притяжательные местоимения    5822  undefined
25992                      ирина ефремова   12653  undefined
12830                     работа в гомеле   22380  undefined
93400         министерство образования рб    3886  undefined
644                   ок ru одноклассники  265034  undefined
3280                      нижний новгород   69315  undefined
21292                            серпухов   14580  undefined
93411               соник икс мультсериал    3877  undefined
53159  казанский собор в санкт петербурге    6479  undefined

### Проверка количества поисковых запросов содержащих города регионов

In [48]:
# делаем проверку чтобы найти поисковые запросы с указанием городов регионов

keywords.groupby('region').count().reset_index()[['region', 'keyword']]

region  keyword
0       undefined    99406
1  Дальний Восток       64
2    Северо-Запад      146
3           Центр      384

### Задание 10

Есть мнение, что “раньше снимали настоящее кино, не то что сейчас”. Ваша задача проверить это утверждение, используя файлы с рейтингами фильмов из прошлого домашнего занятия (файл ratings.csv из базы https://grouplens.org/datasets/movielens). Т. е. проверить верно ли, что с ростом года выпуска фильма его средний рейтинг становится ниже.

При этом мы не будем затрагивать субьективные факторы выставления этих рейтингов, а пройдемся по следующему алгоритму:

В переменную years запишите список из всех годов с 1950 по 2010.

Напишите функцию production_year, которая каждой строке из названия фильма выставляет год выпуска. Не все названия фильмов содержат год выпуска в одинаковом формате, поэтому используйте следующий алгоритм:

для каждой строки пройдите по всем годам списка years
если номер года присутствует в названии фильма, то функция возвращает этот год как год выпуска
если ни один из номеров года списка years не встретился в названии фильма, то возвращается 1900 год
Запишите год выпуска фильма по алгоритму пункта 2 в новый столбец ‘year’

Посчитайте средний рейтинг всех фильмов для каждого значения столбца ‘year’ и отсортируйте результат по убыванию рейтинга

In [49]:
# датафрейм с фильмами

tab_movies = pd.read_csv('../ml-25m/movies.csv')
tab_movies.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [50]:
# список годов от 1950 по 2010

years = list(range(1950, 2011))

In [51]:
# функция ищет год в названии фильма

def production_year(title):
    for year in years:  #через цикл проверяем каждый год на наличие в названии фильма
        if str(year) in title:
            return int(year)
    return 1900

In [52]:
# применяем функцию к датафрейму

tab_movies['year'] = tab_movies['title'].apply(production_year)
tab_movies.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  year  
0  Adventure|Animation|Children|Comedy|Fantasy  1995  
1                   Adventure|Children|Fantasy  1995  
2                               Comedy|Romance  1995  
3                         Comedy|Drama|Romance  1995  
4                                       Comedy  1995

In [53]:
# фильтруем нужные нам столбцы с учёт периода с 1950 по 2010 года

movies_year = tab_movies[['movieId', 'title', 'year']].query('year >= 1950 and year <= 2010')
movies_year.head()

movieId                               title  year
0        1                    Toy Story (1995)  1995
1        2                      Jumanji (1995)  1995
2        3             Grumpier Old Men (1995)  1995
3        4            Waiting to Exhale (1995)  1995
4        5  Father of the Bride Part II (1995)  1995

In [54]:
average_rating.head()

rating            class
movieId                         
1          3.89  Средний рейтинг
2          3.25  Средний рейтинг
3          3.14  Средний рейтинг
4          2.85  Средний рейтинг
5          3.06  Средний рейтинг

In [55]:
# соединяем два датафрейма в один отфильтрованные по годам фильмы и их средний рейтинг, сортируем по убыванию рейтинга

joined = movies_year.merge(average_rating, on='movieId', how='left')
joined.sort_values('rating', ascending=False).head(15)

movieId                                              title  year  \
29687   164620                               Zig Zag Story (1983)  1983   
19839   124482                          The Heartbreak Kid (1993)  1993   
32576   178217                   Hempsters: Plant the Seed (2008)  2008   
19640   123771                           The Cariboo Trail (1950)  1950   
32608   178417                               A Lustful Man (1961)  1961   
8726     27914  Hijacking Catastrophe: 9/11, Fear & the Sellin...  2004   
36396   199762                        Der letzte Fußgänger (1960)  1960   
19701   123959                                     Groupie (2010)  2010   
36393   199756                    Zum Teufel mit der Penne (1968)  1968   
32671   178739                                Kiss Me Kate (2003)  2003   
19751   124174                             Until They Sail (1957)  1957   
24581   143422                                           2 (2007)  2007   
24545   143269                           Moro Witch Doctor (1964)  1964   
24539   143235                            Going to the Mat (2004)  2004   
24535   143223      It Seemed Like a Good Idea at the Time (1975)  1975   

       rating            class  
29687     5.0  Высокий рейтинг  
19839     5.0  Высокий рейтинг  
32576     5.0  Высокий рейтинг  
19640     5.0  Высокий рейтинг  
32608     5.0  Высокий рейтинг  
8726      5.0  Высокий рейтинг  
36396     5.0  Высокий рейтинг  
19701     5.0  Высокий рейтинг  
36393     5.0  Высокий рейтинг  
32671     5.0  Высокий рейтинг  
19751     5.0  Высокий рейтинг  
24581     5.0  Высокий рейтинг  
24545     5.0  Высокий рейтинг  
24539     5.0  Высокий рейтинг  
24535     5.0  Высокий рейтинг

## Проверка нашей гипотезы:
Опираясь на вывод первых 15 строк,в нашем веке также есть достойные и стоящие просмотра фильмы. 

In [56]:
from pymystem3 import Mystem

### Задание 11

Для датафрейма log из материалов занятия создайте столбец source_type по следующим правилам:

- если источник traffic_source равен yandex или google, то в source_type ставится organic
- для источников paid и email из России - ставим ad
- для источников paid и email не из России - ставим other
- все остальные варианты берем из traffic_source без изменений

In [57]:
import os

### Решение

In [61]:
df_log = pd.read_csv(os.path.join('C:', os.sep, 'Netology', 'Python_12_pivot_and_str', 'visit_log.csv'), sep=';')

In [59]:
df_log.loc[(df_log.traffic_source == 'yandex') | (df_log.traffic_source == 'google'), 'source_type'] = 'organic'
df_log.loc[df_log.traffic_source.isin(['paid', 'email']) & (df_log.region == 'Russia'), 'source_type'] = 'ad'
df_log.loc[(df_log.traffic_source.isin(['paid', 'email'])) & (df_log.region != 'Russia'), 'source_type'] = 'other'
df_log.loc[(~df_log.traffic_source.isin(['yandex', 'google', 'paid', 'email'])) & ((df_log.region != 'Russia') | (df_log.region.isin(['Russia']))), 'source_type'] = df_log['traffic_source']

In [60]:
df_log.head(20)

timestamp    visit_id                                 url   region  \
0   1549980692  e3b0c44298  https://host.ru/3c19b4ef7371864fa3   Russia   
1   1549980704  6e340b9cff  https://host.ru/c8d9213a31839f9a3a   Russia   
2   1549980715  96a296d224  https://host.ru/b8b58337d272ee7b15   Russia   
3   1549980725  709e80c884  https://host.ru/b8b58337d272ee7b15   Russia   
4   1549980736  df3f619804  https://host.ru/b8b58337d272ee7b15   Russia   
5   1549980742  8855508aad  https://host.ru/df646c3676cc259fa0   Russia   
6   1549980742  b0f66adc83  https://host.ru/b8b58337d272ee7b15   Russia   
7   1549980754  837885c8f8  https://host.ru/108ce4b365afb7b88e   Russia   
8   1549980760  af5570f5a1  https://host.ru/3004a8273caeef2867    China   
9   1549980765  3e7077fd2f  https://host.ru/df646c3676cc259fa0   Russia   
10  1549980777  01d448afd9  https://host.ru/a5dda93e70318570c0    India   
11  1549980785  71b6c1d538  https://host.ru/c8d9213a31839f9a3a  Ukraine   
12  1549980789  15ec7bf0b5  https://host.ru/8e88d4703848cc0ec4   Russia   
13  1549980792  dd46c3eebb  https://host.ru/3004a8273caeef2867    China   
14  1549980793  e7ecebbc59  https://host.ru/05e41944dd85a900a6  Belarus   
15  1549980797  5322fecfc9  https://host.ru/4adc8a29b212051b04  Belarus   
16  1549980801  374708fff7  https://host.ru/39fa884393666d45fc   Russia   
17  1549980807  0a88111852  https://host.ru/2fcb223266880ccf57  Ukraine   
18  1549980814  60daa3a5f7  https://host.ru/ef57543002ce2a0f2a  Belarus   
19  1549980816  d6fd62f5ce  https://host.ru/f1eb4601740d627ab0  Belarus   

       user_id traffic_source source_type  
0   b1613cc09f         yandex     organic  
1   4c3ec14bee         direct      direct  
2   a8c40697fb         yandex     organic  
3   521ac1d6a0         yandex     organic  
4   d7323c571c         yandex     organic  
5   fc43898e47         yandex     organic  
6   13fc55e781           paid          ad  
7   cb5082b6f6         direct      direct  
8   45664f7af2         direct      direct  
9   6f9de8c8b6          email          ad  
10  27e76a4a80         direct      direct  
11  3b13701488         google     organic  
12  6d1d81f7f8           paid          ad  
13  7e301c8c41         direct      direct  
14  ac52dedc3b         direct      direct  
15  9a45d98867           paid       other  
16  e9bd59e22d         google     organic  
17  a2ba8fa3db           paid       other  
18  72cb10fb5c         direct      direct  
19  7269dcd869         direct      direct

In [62]:
## Альтернативное решение
df_log['source_type'] = df_log['traffic_source']
df_log.loc[df_log.traffic_source.isin(['yandex', 'google']), 'source_type'] = 'organic'
df_log.loc[df_log.traffic_source.isin(['paid', 'email']) & df_log.region.isin(['Russia']), 'source_type'] = 'ad'
df_log.loc[(df_log.traffic_source.isin(['paid', 'email'])) & (df_log.region != 'Russia'), 'source_type'] = 'other'
df_log.head(20)

timestamp    visit_id                                 url   region  \
0   1549980692  e3b0c44298  https://host.ru/3c19b4ef7371864fa3   Russia   
1   1549980704  6e340b9cff  https://host.ru/c8d9213a31839f9a3a   Russia   
2   1549980715  96a296d224  https://host.ru/b8b58337d272ee7b15   Russia   
3   1549980725  709e80c884  https://host.ru/b8b58337d272ee7b15   Russia   
4   1549980736  df3f619804  https://host.ru/b8b58337d272ee7b15   Russia   
5   1549980742  8855508aad  https://host.ru/df646c3676cc259fa0   Russia   
6   1549980742  b0f66adc83  https://host.ru/b8b58337d272ee7b15   Russia   
7   1549980754  837885c8f8  https://host.ru/108ce4b365afb7b88e   Russia   
8   1549980760  af5570f5a1  https://host.ru/3004a8273caeef2867    China   
9   1549980765  3e7077fd2f  https://host.ru/df646c3676cc259fa0   Russia   
10  1549980777  01d448afd9  https://host.ru/a5dda93e70318570c0    India   
11  1549980785  71b6c1d538  https://host.ru/c8d9213a31839f9a3a  Ukraine   
12  1549980789  15ec7bf0b5  https://host.ru/8e88d4703848cc0ec4   Russia   
13  1549980792  dd46c3eebb  https://host.ru/3004a8273caeef2867    China   
14  1549980793  e7ecebbc59  https://host.ru/05e41944dd85a900a6  Belarus   
15  1549980797  5322fecfc9  https://host.ru/4adc8a29b212051b04  Belarus   
16  1549980801  374708fff7  https://host.ru/39fa884393666d45fc   Russia   
17  1549980807  0a88111852  https://host.ru/2fcb223266880ccf57  Ukraine   
18  1549980814  60daa3a5f7  https://host.ru/ef57543002ce2a0f2a  Belarus   
19  1549980816  d6fd62f5ce  https://host.ru/f1eb4601740d627ab0  Belarus   

       user_id traffic_source source_type  
0   b1613cc09f         yandex     organic  
1   4c3ec14bee         direct      direct  
2   a8c40697fb         yandex     organic  
3   521ac1d6a0         yandex     organic  
4   d7323c571c         yandex     organic  
5   fc43898e47         yandex     organic  
6   13fc55e781           paid          ad  
7   cb5082b6f6         direct      direct  
8   45664f7af2         direct      direct  
9   6f9de8c8b6          email          ad  
10  27e76a4a80         direct      direct  
11  3b13701488         google     organic  
12  6d1d81f7f8           paid          ad  
13  7e301c8c41         direct      direct  
14  ac52dedc3b         direct      direct  
15  9a45d98867           paid       other  
16  e9bd59e22d         google     organic  
17  a2ba8fa3db           paid       other  
18  72cb10fb5c         direct      direct  
19  7269dcd869         direct      direct

### Задание 12
В файле URLs.txt содержатся url страниц новостного сайта. Вам необходимо отфильтровать его по адресам страниц с текстами новостей. Известно, что шаблон страницы новостей имеет внутри url следующую конструкцию: /, затем 8 цифр, затем дефис. Выполните следующие действия:

Прочитайте содержимое файла с датафрейм
Отфильтруйте страницы с текстом новостей, используя метод str.contains и регулярное выражение в соответствии с заданным шаблоном

### Решение

In [64]:
file_url = df_log = pd.read_csv(os.path.join('C:', os.sep, 'Netology', 'Python_12_pivot_and_str', 'URLs.txt'))
file_url.head()

url
0                                            /world/
1                                           /latest/
2                                      /?updated=top
3  /politics/36188461-s-marta-zhizn-rossiyan-susc...
4  /world/36007585-tramp-pridumal-kak-reshit-ukra...

In [65]:
import re

file_url[file_url.url.str.contains(r'/[0-9]{8}-')].head()

url
3  /politics/36188461-s-marta-zhizn-rossiyan-susc...
4  /world/36007585-tramp-pridumal-kak-reshit-ukra...
5  /science/36157853-nasa-sobiraet-ekstrennuyu-pr...
6  /video/36001498-poyavilis-pervye-podrobnosti-g...
7  /world/36007585-tramp-pridumal-kak-reshit-ukra...

### Задание 13
В датафрейме data создайте столбец lemmas, в котором вычислите леммы поисковых запросов из столбца keyword. Леммы должны иметь строковый тип.

### Решение

In [66]:
data = pd.DataFrame({
    'keyword': ['курс гривны к рублю', 'доллары в рубли', '100 долларов в рублях', 'курс рубля'],
    'shows': [125076, 114173, 97534, 53546],
})

In [67]:
def lemmas(keyword):
    m = Mystem()
    lemmas = m.lemmatize(keyword)
    return ''.join(lemmas).strip()
    

In [68]:
data['lemmas'] = data['keyword'].apply(lemmas)
data

keyword   shows               lemmas
0    курс гривны к рублю  125076  курс гривна к рубль
1        доллары в рубли  114173       доллар в рубль
2  100 долларов в рублях   97534   100 доллар в рубль
3             курс рубля   53546           курс рубль

### Задание 14
Используйте файл с оценками фильмов ml-latest-small/ratings.csv. Посчитайте среднее время жизни пользователей, которые выставили более 100 оценок. Под временем жизни понимается разница между максимальным и минимальным значением столбца timestamp для данного значения userId.

### Решение

In [69]:
ratings = pd.read_csv(os.path.join('C:', os.sep, 'Netology', 'ml-latest-small', 'ratings.csv'))
ratings.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [70]:
filtered_ratings = ratings.groupby('userId').agg({'timestamp': ['min', 'max', 'count']})['timestamp'].query('count > 100').reset_index()
filtered_ratings = filtered_ratings[['userId', 'min', 'max', 'count']]
filtered_ratings.head()

userId         min         max  count
0       4   949778714   949982274    204
1       8  1154389340  1154474527    116
2      15   997937239  1469330735   1700
3      17  1127468587  1127476640    363
4      19   855190091   855195373    423

In [71]:
filtered_ratings['lifetime'] = filtered_ratings['max'] - filtered_ratings['min'] 
filtered_ratings.head()

userId         min         max  count   lifetime
0       4   949778714   949982274    204     203560
1       8  1154389340  1154474527    116      85187
2      15   997937239  1469330735   1700  471393496
3      17  1127468587  1127476640    363       8053
4      19   855190091   855195373    423       5282

In [72]:
day = 86400 # 1 день равен 86400 секунд

mean_lifetime = round((sum(filtered_ratings['lifetime']) / filtered_ratings['userId'].count()) / day, 0)
mean_lifetime

464.0

Ответ:
Среднее время жизни пользователей, которые выставили более 100 оценок равно 464 дня

### Задание 15

Дана статистика услуг перевозок клиентов компании по типам (см. файл с кодом занятия). Необходимо сформировать две таблицы:
таблицу с тремя типами выручки для каждого client_id без указания адреса клиента
аналогичную таблицу по типам выручки с указанием адреса клиента
Обратите внимание, что в процессе объединения таблиц данные не должны теряться.

Дана статистика услуг перевозок клиентов компании по типам:

- rzd - железнодорожные перевозки
- auto - автомобильные перевозки
- air - воздушные перевозки
- client_base - адреса клиентов

### Решение

In [73]:
rzd = pd.DataFrame(
    {
        'client_id': [111, 112, 113, 114, 115],
        'rzd_revenue': [1093, 2810, 10283, 5774, 981]
    }
)
rzd

client_id  rzd_revenue
0        111         1093
1        112         2810
2        113        10283
3        114         5774
4        115          981

In [74]:
auto = pd.DataFrame(
    {
        'client_id': [113, 114, 115, 116, 117],
        'auto_revenue': [57483, 83, 912, 4834, 98]
    }
)
auto

client_id  auto_revenue
0        113         57483
1        114            83
2        115           912
3        116          4834
4        117            98

In [75]:
air = pd.DataFrame(
    {
        'client_id': [115, 116, 117, 118],
        'air_revenue': [81, 4, 13, 173]
    }
)
air

client_id  air_revenue
0        115           81
1        116            4
2        117           13
3        118          173

In [76]:
client_base = pd.DataFrame(
    {
        'client_id': [111, 112, 113, 114, 115, 116, 117, 118],
        'address': ['Комсомольская 4', 'Энтузиастов 8а', 'Левобережная 1а', 'Мира 14', 'ЗЖБИиДК 1', 
                    'Строителей 18', 'Панфиловская 33', 'Мастеркова 4']
    }
)
client_base

client_id          address
0        111  Комсомольская 4
1        112   Энтузиастов 8а
2        113  Левобережная 1а
3        114          Мира 14
4        115        ЗЖБИиДК 1
5        116    Строителей 18
6        117  Панфиловская 33
7        118     Мастеркова 4

In [77]:
joined_client_1 = (
    client_base.filter( items = ['client_id'])
    .merge(rzd, on='client_id', how='left')
    .merge(auto, on='client_id', how='left')
    .merge(air, on='client_id', how='left')
)
joined_client_1.fillna(0).head(10)

client_id  rzd_revenue  auto_revenue  air_revenue
0        111       1093.0           0.0          0.0
1        112       2810.0           0.0          0.0
2        113      10283.0       57483.0          0.0
3        114       5774.0          83.0          0.0
4        115        981.0         912.0         81.0
5        116          0.0        4834.0          4.0
6        117          0.0          98.0         13.0
7        118          0.0           0.0        173.0

In [78]:
joined_client_2 = (
    client_base
    .merge(rzd, on='client_id', how='left')
    .merge(auto, on='client_id', how='left')
    .merge(air, on='client_id', how='left')
)
joined_client_2.fillna(0).head(10)

client_id          address  rzd_revenue  auto_revenue  air_revenue
0        111  Комсомольская 4       1093.0           0.0          0.0
1        112   Энтузиастов 8а       2810.0           0.0          0.0
2        113  Левобережная 1а      10283.0       57483.0          0.0
3        114          Мира 14       5774.0          83.0          0.0
4        115        ЗЖБИиДК 1        981.0         912.0         81.0
5        116    Строителей 18          0.0        4834.0          4.0
6        117  Панфиловская 33          0.0          98.0         13.0
7        118     Мастеркова 4          0.0           0.0        173.0

### Задание 16
В задаче сквозной аналитики вам предоставили данные по местоположению пользователей. Т. е. для каждого user_id известна последовательность координат (широта/долгота), когда они требовались приложению для полноценной работы. Как бы вы добавили эти сведения в таблицу визитов и покупок? Для составления ответа можно использовать вопросы:

- У каждого пользователя известен набор координат. А для связывания с визитом или фактом покупки скорее всего потребуется одно-два числа. Как их получить?
- Наборы координат одного и того же пользователя могут быть значительно удалены друг от друга. Как это отразится на вопросе расчетах пункта 1?
- Какие дополнительные признаки можно получить из координат? Ведь это просто числа, которые сами по себе мало что дают.

### Размышления:

Координаты являются ункальными значениями, один адрес пользователя, одно его местоположение это две координаты, у пользователя может быть один или несколько адресов, то есть при добавлении координат в таблица визитов и покупок, по итогу мы должны получить строку: user_id, визиты, покупки, адрес(или координаты), где user_id уникален, но может повторяться в строках (в зависимости от количества адресов), визиты и покупки мы можем суммировать, возможно мы можем разместить в строке один адрес, а можем список адресов одного пользователя, тогда строка будет выглядеть так: user_id, sum(визитов), sum(покупок), список адресов, то есть допускаем и тот и тот вариант, в зависимости от наших целей.

Если у пользователя несколько местоположений, то в таком случаем из координат нескольких местоположений пользователя, можно составить полигон, вершинами которого будут координаты местоположений пользователя, тогда наша строка будет как: user_id, sum(визитов), sum(покупок), полигон.

Связывание набора координат и визиты и покупки пользователя, будет через user_id, который у каждого пользователя уникален.
Иначе не совсем понятен вопрос, о какой взаимосвязи речь, о той которая поможет объединить данные(таблицы), или как связаны визиты и покупки с координатами (думается, что никак)?

При расчете пункта один, если у пользователя несколько местоположение, то либо мы получим полигон пользователя, и сумму визитов и покупок и его уникальный user_id, 
или строки где каждому пользователю будет пренадлежать несколько строк с данными по визитам и покупкам и его уникальным адресам, если с одного адреса происходило несколько визитов и покупок, то в таком случаем можно сгруппировать данные (по принципу суммы).

### Задание 17

Обязательная часть
Написать скрипт, который будет проверять список e-mail адресов на утечку при помощи сервиса Avast Hack Ckeck. Список email-ов задаем переменной в начале кода:
EMAIL = [xxx@x.ru, yyy@y.com]

В итоге должен формироваться датафрейм со столбцами: <почта> - <дата утечки> - <источник утечки> - <описание утечки>

Подсказка: сервис работает при помощи "скрытого" API. Внимательно изучите post-запросы.

### Решение

In [88]:
import json
import requests

# main part

EMAIL = ['xxx@x.ru', 'yyy@y.com']
url = 'https://identityprotection.avast.com/v1/web/query/site-breaches/unauthorized-data'
headers = {'Vaar-Header-App-Build-Version': '1.0.0', 'Vaar-Header-App-Product': 'hackcheck-web-avast', 'Vaar-Header-App-Product-Name': 'hackcheck-web-avast', 'Vaar-Version': '0'}
adresses = {'emailAddresses': EMAIL}
adresses = {'emailAddresses': EMAIL}
res = requests.post(url, json=adresses, headers=headers)
df = pd.DataFrame(res.json()['breaches']).T.reset_index() 
df = df[['site', 'description', 'publishDate']]

In [89]:
df.head()

site                                        description  \
0         azcentral.com  At an unconfirmed date, online Arizona newspap...   
1           wishbone.io  In January 2020, the online poll website Wishb...   
2        myheritage.com  In October 2017, a customer database belonging...   
3  forums.vkmonline.com  At an unconfirmed date, the Russian-language m...   
4             canva.com  In May 2019, graphic-design site Canva's datab...   

            publishDate  
0  2020-01-03T00:00:00Z  
1  2020-05-28T00:00:00Z  
2  2017-11-04T00:00:00Z  
3  2021-02-11T00:00:00Z  
4  2019-06-13T00:00:00Z